_posts/2020-06-28-openfaas-loki-into.md

```diff
@@ -0,0 +1,235 @@
+---
+title: Time travelling through your OpenFaaS Function logs
+description: "In this post Lucas will introduce the openfaas-loki log provider and how to use it to get historic logs for your OpenFaaS functions."
+date: 2020-05-28
+image: /images/2020-openfaas-loki/2020-06-loki-logs-timey-wobbly.png
+categories:
+  - logs
+  - kubernetes
+  - serverless
+  - tutorial
+  - grafana
+  - loki
+author_staff_member: lucas
+dark_background: true
+
+---
+
+In this post Lucas will introduce the [`openfaas-loki`](https://github.com/LucasRoesler/openfaas-loki) [log provider](https://docs.openfaas.com/architecture/logs-provider/) and how to use it to get historic logs for your OpenFaaS functions.
+
+In any production service, logs are a critical component of observability and debuggability of the system. OpenFaaS builds in support for logs in _every_ installation. But it can be even better by using a new log provider. In this post we will show you how to supercharge the logs by installing [Grafana's Loki](https://grafana.com/oss/loki) and configure OpenFaaS to use Loki.
+
+## Motivation
+
+In the default installation of OpenFaaS, the logs come from the underlying orchestration tool. This is great, because logs are automatically available without any additional effort, but this means logs only exist for the _current_ function instances. This means that as your function scales from 0 to 100 and then back to 0, you will no long have access to the function logs because the orchestraition system does not keep logs after the function instance is removed! This is simply the default behavior of systems like Docker Swarm and Kubernetes.  If a user reports and error, but the function instance has scaled down already, the logs you need to debug the issue are now gone :(
+
+To have logs for long ago terminated function instances, we need to install some kind of logs aggregator. Traditionally, this has meant the [ELK stack](https://www.elastic.co/what-is/elk-stack): ElasticSearch, Logstash, Kibana. This is a great and well established tool; but, in [late 2018 Grafana introduced](https://kccna18.sched.com/event/GrXC/on-the-oss-path-to-full-observability-with-grafana-david-kaltschmidt-grafana-labs) a new alternative aimed squarely at the Cloud Native world: [Loki](https://grafana.com/oss/loki/).
```

Comment:
> To have logs for long ago terminated function instances, we need to install some kind of logs aggregator

I think the beginning of the sentence this sounds a bit awkward, would this work?

> Sometimes it's important to retrieve the logs for function invocations from several days or even weeks ago. To achieve that, we need to aggregate, store and index the logs from each function.